rosetta/retriever/retriever_test.go

```diff
@@ -0,0 +1,1066 @@
+				Identifier: "address",
+				Type:       cadence.AddressType{},
+			},
+		},
+	}
+	depositEvent := cadence.NewEvent(
+		[]cadence.Value{
+			cadence.NewUInt64(42),
+			cadence.NewAddress([8]byte{1, 2, 3, 4, 5, 6, 7, 8}),
+		},
+	).WithType(depositType)
+	depositEventPayload := json.MustEncode(depositEvent)
+
+	withdrawalType := &cadence.EventType{
+		Location:            utils.TestLocation,
+		QualifiedIdentifier: "withdrawal",
+		Fields: []cadence.Field{
+			{
+				Identifier: "amount",
+				Type:       cadence.UInt64Type{},
+			},
+			{
+				Identifier: "address",
+				Type:       cadence.AddressType{},
+			},
+		},
+	}
+	withdrawalEvent := cadence.NewEvent(
+		[]cadence.Value{
+			cadence.NewUInt64(42),
+			cadence.NewAddress([8]byte{2, 3, 4, 5, 6, 7, 8, 9}),
+		},
+	).WithType(withdrawalType)
+	withdrawalEventPayload := json.MustEncode(withdrawalEvent)
+
+	testEvents := []flow.Event{
+		{
+			Type:          "deposit",
+			TransactionID: id,
+			EventIndex:    0,
+			Payload:       depositEventPayload,
+		},
+		{
+			TransactionID: id,
+			Type:          "withdrawal",
+			EventIndex:    1,
+			Payload:       withdrawalEventPayload,
+		},
+	}
+
+	t.Run("nominal case", func(t *testing.T) {
+		index := &mocks.Reader{
+			HeaderFunc: func(height uint64) (*flow.Header, error) {
+				assert.Equal(t, testHeight, height)
+				return testHeader, nil
+			},
+			EventsFunc: func(height uint64, types ...flow.EventType) ([]flow.Event, error) {
+				assert.Equal(t, testHeight, height)
+				if assert.Len(t, types, 2) {
+					assert.Equal(t, flow.EventType("deposit"), types[0])
+					assert.Equal(t, flow.EventType("withdrawal"), types[1])
+				}
+				return testEvents, nil
+			},
+		}
+		validator := &mocks.Validator{
+			BlockFunc: func(block identifier.Block) (identifier.Block, error) {
+				assert.Equal(t, testBlockID, block)
+				return block, nil
+			},
+		}
+		generator := &mocks.Generator{
+			TokensDepositedFunc: func(symbol string) (string, error) {
+				assert.Equal(t, symbol, dps.FlowSymbol)
+				return "deposit", nil
+			},
+			TokensWithdrawnFunc: func(symbol string) (string, error) {
+				assert.Equal(t, symbol, dps.FlowSymbol)
+				return "withdrawal", nil
+			},
+		}
+
+		r := &Retriever{
+			index:     index,
+			validate:  validator,
+			generator: generator,
+		}
+
+		// TODO: Add verification for transactions when https://github.com/optakt/flow-dps/issues/149 is implemented.
+		block, _, err := r.Block(testBlockID)
+
+		if assert.NoError(t, err) {
+			assert.Equal(t, testBlock, block)
+		}
+	})
+
+	t.Run("handles invalid block", func(t *testing.T) {
+		validator := &mocks.Validator{
+			BlockFunc: func(block identifier.Block) (identifier.Block, error) {
+				assert.Equal(t, testBlockID, block)
+				return identifier.Block{}, mocks.DummyError
+			},
+		}
+
+		r := &Retriever{
+			validate: validator,
+		}
+
+		_, _, err := r.Block(testBlockID)
+		assert.Error(t, err)
+	})
+
+	t.Run("handles generator failures", func(t *testing.T) {
```

Comment:
Maybe we could qualify these two test cases with their own names?

`handle generator failures (deposit)` and `handle generator failures (withdrawal)`?